File: Algorithm/src/slidingwindow/H8_LongestSubarraywithOnesafterReplacement.java
package slidingwindow;

public class H8_LongestSubarraywithOnesafterReplacement {
  public static int find(int[] arr, int k){
    int start = 0, max = 0, zero = 0, one = 0;

    for(int end = 0; end < arr.length; end++){
      int cur = arr[end];
      if(cur == 0){
        zero++;
      }else{
        one++;
      }
      while(zero > k){
        int reduce = arr[start];
        if(reduce == 0){
          zero--;
        }else{
          one--;
        }
        start++;
      }
      max = Math.max(max, end - start + 1);
    }
    return max;
  }

  public static void main(String[] args) {
    System.out.println(find(new int[] {0,1,1,0,0,0,0,1,1,0,1,1}, 2));
    System.out.println(find(new int[] {0,1,0,0,1,1,0,1,1,0,0,1,1}, 3));
  }
}

File: Algorithm/src/slidingwindow/H8_LongestSubarraywithOnesafterReplacement.py
def find(arr, k):
  start, maxlen, count = 0, 0, 0
  for end in range(0, len(arr)):
    if arr[end] == 1:
      count += 1
    while end - start + 1 - count > k:
      if arr[start] == 1:
        count -= 1
      start += 1
    maxlen = max(maxlen, end - start + 1)
  return maxlen

def main():
  print(find([0,1,1,0,0,0,0,1,1,0,1,1], 2))

main()